equinox/nn/_kv_cache.py

```diff
@@ -0,0 +1,103 @@
+from abc import abstractmethod
+
+import jax.lax as lax
+import jax.numpy as jnp
+from jaxtyping import Array, Float, Int
+
+from .._misc import default_int_dtype
+from .._module import field, Module
+from ._stateful import State, StateIndex
+
+
+class AbstractKVCache(Module, strict=True):
```

Comment:
So I suspect this doesn't need to actually be a full-blown abstract class. It looks like all the caller needs is a `__call__` with the appropriate API.

I think this means that we could have
```python
class MultiheadAttention(eqx.Module:
    def __init__(..., kv_cache: Optional[Callable[...signature here...]] = None):

class StandardKVCache(eqx.Module): ...
```
with the latter just collapsing the existing implementation of `AbstractKVCache` and `StandardKVCache` into just one class.

For bonus points this would mean we achieve consistency with the API for `process_heads`, which also just uses a `Callable` rather than an ABC. (In fact I don't think we really have ABCs anywhere else in `eqx.nn` -- neural networks seem to be simply enough that they're not needed.)

Reply:
As a side note, some comment on the design as you currently have it (just for your own interest):
- `AbstractKVCache` is defining an `autoregressive_index` attribute, but this isn't actually used anywhere that `AbstractKVCache` is accepted. Rather, it is only used inside the `StandardKVCache` subclass. So this should be moved into the subclass.
- You might [this doc](https://docs.kidger.site/equinox/pattern/) interesting! It's the pattern I'm trying to use in Equinox and its ecosystem going forward. In particualr it advocates against `super`.